enterprise/tools/upload_local_snapshot/upload_local_snapshot.go

```diff
@@ -0,0 +1,274 @@
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/nullauth"
+	"github.com/buildbuddy-io/buildbuddy/server/real_environment"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/cachetools"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
+	"github.com/buildbuddy-io/buildbuddy/server/util/flagutil"
+	"github.com/buildbuddy-io/buildbuddy/server/util/grpc_client"
+	"github.com/buildbuddy-io/buildbuddy/server/util/hash"
+	"github.com/buildbuddy-io/buildbuddy/server/util/healthcheck"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/proto"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	fcpb "github.com/buildbuddy-io/buildbuddy/proto/firecracker"
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	rspb "github.com/buildbuddy-io/buildbuddy/proto/resource"
+	bspb "google.golang.org/genproto/googleapis/bytestream"
+)
+
+// This tool should be run on the executor with the desired local snapshot in
+// its filecache. It will upload the local snapshot to the remote cache,
+// so it can be more easily inspected and run with vmstart, for example.
+
+var (
+	cacheTarget  = flag.String("cache_target", "grpcs://remote.buildbuddy.dev", "The remote cache target to upload the snapshot to.")
+	filecacheDir = flag.String("file_cache_dir", "/tmp/buildbuddy/filecache", "The local filecache dir on the executor")
+	apiKey       = flag.String("api_key", "", "The API key to use to interact with the remote cache.")
+	// Note: this key can be copied and pasted from logs:
+	//
+	//  "INFO: Found snapshot for key {...}" // copy this JSON
+	//
+	// At the shell, paste it in single quotes: -remote_snapshot_key='<paste>'
+	remoteSnapshotKeyJSON = flag.String("remote_snapshot_key", "", "JSON struct containing a local snapshot key that should be uploaded to the remote cache.")
+)
+
+type RemoteSnapshotKey struct {
+	GroupID      string `json:"group_id"`
+	InstanceName string `json:"instance_name"`
+	KeyDigest    string `json:"key_digest"`
+	Key          any    `json:"key"`
+}
+
+func main() {
+	flag.Parse()
+
+	if *remoteSnapshotKeyJSON == "" {
+		log.Fatalf("You must pass in a snapshot key")
+	}
+	inputKey, key, err := parseSnapshotKeyJSON(*remoteSnapshotKeyJSON)
+	if err != nil {
+		log.Fatalf("Failed to parse snapshot key: %s", err)
+	}
+	snapshotGroupID := inputKey.GroupID
+	snapshotInstanceName := inputKey.InstanceName
+
+	ctx := context.Background()
+	if *apiKey != "" {
+		ctx = metadata.AppendToOutgoingContext(ctx, "x-buildbuddy-api-key", *apiKey)
+	}
+	env := getToolEnv()
+	loader, err := snaploader.New(env)
+	if err != nil {
+		log.Fatalf("Failed to init snaploader: %s", err)
+	}
+
+	allSnapshotDigests := getAllDigestsFromSnapshotManifest(ctx, env, snapshotInstanceName, snapshotGroupID, key)
```

Comment:
we should probably return the manifest here so that `uploadManifest` uploads the same manifest that we read here (since the manifest stored at the FC key might change while we're uploading)

Reply:
Good call - done